bokeh/models/sources.py

```diff
@@ -171,6 +173,34 @@ def add(self, data, name=None):
         self.data[name] = data
         return name
 
+    def traverse_data(self, datum):
+        datum_copy = []
+        for item in datum:
+            if isinstance(item, (list, tuple)):
+                datum_copy.append(self.traverse_data(item))
+            elif isinstance(item, float):
+                if np.isnan(item):
+                    item = 'NaN'
+                elif np.isposinf(item):
+                    item = 'Infinity'
+                elif np.isneginf(item):
+                    item = '-Infinity'
+                datum_copy.append(item)
+            else:
+                datum_copy.append(item)
+        return datum_copy
+
+    def transform_column_source_data(self, data):
+        data_copy = {}
+        for key in iterkeys(data):
+            data_copy[key] = self.traverse_data(data[key])
+        return data_copy
+
+    def vm_serialize(self, changed_only=True):
+        attrs = super(ColumnDataSource, self).vm_serialize(changed_only=changed_only)
+        self.data.update(self.transform_column_source_data(self.data))
```

Comment:
Isn't this changing the original model? We only want to affect the serialized output.